### PR TITLE
Feat boomerang canister

### PR DESCRIPTION
**Pull request by uk4z**
Adding a testing framework to the Boomerang canister.

### CODE REVIEW

File: boomerang/src/state_machine/canisters/ledger-canister.wasm.gz
Comment (EnzoPlayer0ne) [inline boomerang/src/state_machine/canisters/ledger-canister.wasm.gz:1]: need to remove this gz

File: boomerang/src/state_machine/canisters/governance-canister.wasm.gz
Comment (EnzoPlayer0ne) [inline boomerang/src/state_machine/canisters/governance-canister.wasm.gz:1]: ditto

File: boomerang/src/state_machine/sns_canisters/sns-governance-canister.wasm.gz
Comment (EnzoPlayer0ne) [inline boomerang/src/state_machine/sns_canisters/sns-governance-canister.wasm.gz:1]: same here

File: boomerang/src/state_machine/canisters/ledger-canister.wasm.gz
Comment (uk4z) [reply boomerang/src/state_machine/canisters/ledger-canister.wasm.gz:1]: done

File: boomerang/src/state_machine/canisters/governance-canister.wasm.gz
Comment (uk4z) [reply boomerang/src/state_machine/canisters/governance-canister.wasm.gz:1]: done

File: boomerang/src/state_machine/sns_canisters/sns-governance-canister.wasm.gz
Comment (uk4z) [reply boomerang/src/state_machine/sns_canisters/sns-governance-canister.wasm.gz:1]: done

**Review by EnzoPlayer0ne — APPROVED** [review]

File: Cargo.Bazel.lock [hunk 1]
```diff
@@ -8973,6 +8973,10 @@
               "id": "ic-icrc1-ledger 0.9.0",
               "target": "ic_icrc1_ledger"
```
Comment (uk4z) [inline Cargo.Bazel.lock:8974]: test

File: boomerang/src/lib.rs [hunk 2]
```diff
@@ -9,17 +9,32 @@ pub mod icp_to_nicp;
...
+// "renrk-eyaaa-aaaaa-aaada-cai"
+#[cfg(feature = "test-env")]
+pub const WTN_LEDGER_ID: Principal = Principal::from_slice(&[0, 0, 0, 0, 0, 0, 0, 6, 1, 1]);
 
 // "tsbvt-pyaaa-aaaar-qafva-cai"
+#[cfg(not(feature = "test-env"))]
 pub const WATER_NEURON_ID: Principal = Principal::from_slice(&[0, 0, 0, 0, 2, 48, 1, 106, 1, 1]);
+// "r7inp-6aaaa-aaaaa-aaabq-cai"
+#[cfg(feature = "test-env")]
+pub const WATER_NEURON_ID: Principal = Principal::from_slice(&[0, 0, 0, 0, 0, 0, 0, 3, 1, 1]);
```
Comment (leo-ship-it) [inline boomerang/src/lib.rs:38]: That's not the way to go if you want to go on this path create a state, but if we want to stay light you should create canisters at right ids in state machine, I think there is a method to do that but we have to check.

File: boomerang/src/lib.rs [hunk 3]
```diff
@@ -57,24 +72,24 @@ fn check_canister_ids() {
...
     );
     assert_eq!(
         Principal::from_text("jcmow-hyaaa-aaaaq-aadlq-cai").unwrap(),
-        WTN_LEDGER_ID
+        Principal::from_slice(&[0, 0, 0, 0, 2, 0, 0, 215, 1, 1])
     );
     assert_eq!(
         Principal::from_text("tsbvt-pyaaa-aaaar-qafva-cai").unwrap(),
-        WATER_NEURON_ID
+        Principal::from_slice(&[0, 0, 0, 0, 2, 48, 1, 106, 1, 1])
```
Comment (leo-ship-it) [inline boomerang/src/lib.rs:83]: useless test, you shouldn't remove this test.

File: boomerang/src/lib.rs [hunk 4]
```diff
@@ -57,24 +72,24 @@ fn check_canister_ids() {
...
     );
 }
 
-#[cfg(target = "wasm32-unknown-unkown")]
+#[cfg(target_arch = "wasm32")]
```
Comment (leo-ship-it) [inline boomerang/src/lib.rs:87]: why?

File: boomerang/src/nicp_to_icp.rs [hunk 5]
```diff
@@ -11,10 +11,11 @@ use icrc_ledger_types::icrc1::transfer::TransferArg;
 use icrc_ledger_types::icrc2::approve::ApproveArgs;
 
 pub async fn notify_nicp_deposit(target: Principal) -> Result<WithdrawalSuccess, BoomerangError> {
+    let boomerang_id = self_canister_id();
     let subaccount = derive_subaccount_unstaking(target);
 
     let boomerang_account = Account {
-        owner: self_canister_id(),
+        owner: boomerang_id,
```
Comment (leo-ship-it) [inline boomerang/src/nicp_to_icp.rs:18]: why do you want to add a line?

File: boomerang/src/nicp_to_icp.rs [hunk 6]
```diff
@@ -26,13 +27,15 @@ pub async fn notify_nicp_deposit(target: Principal) -> Result<WithdrawalSuccess,
     let balance_e8s: u64 = match client.balance_of(boomerang_account).await {
         Ok(balance) => balance.0.try_into().unwrap(),
         Err((code, message)) => {
-            return Err(BoomerangError::GenericError { code, message });
+            return Err(BoomerangError::BalanceOfError(format!(
+                "code: {code} - message: {message}"
+            )));
         }
```
Comment (leo-ship-it) [inline boomerang/src/nicp_to_icp.rs:33]: bad error message, use the generic one

File: boomerang/src/icp_to_nicp.rs [hunk 7]
```diff
@@ -19,16 +19,16 @@ pub async fn retrieve_nicp(target: Principal) -> Result<Nat, BoomerangError> {
...
     };
 
-    let nicp_balance_e8s: u64 = match nicp_client.balance_of(target_account).await {
+    let nicp_balance_e8s: u64 = match nicp_client.balance_of(boomerang_account).await {
         Ok(balance) => balance.0.try_into().unwrap(),
-        Err((code, msg)) => {
+        Err((code, message)) => {
             return Err(BoomerangError::BalanceOfError(format!(
-                "code: {code} - msg: {msg}"
+                "code: {code} - message: {message}"
```
Comment (leo-ship-it) [inline boomerang/src/icp_to_nicp.rs:31]: better to use generic error

File: boomerang/src/nicp_to_icp.rs [hunk 8]
```diff
@@ -56,8 +59,10 @@ pub async fn notify_nicp_deposit(target: Principal) -> Result<WithdrawalSuccess,
                 return Err(BoomerangError::ApproveError(error));
             }
         },
-        Err((code, message)) => return Err(BoomerangError::GenericError { code, message }),
-    };
+        Err((code, message)) => {
+            return Err(BoomerangError::GenericError { code, message });
+        }
```
Comment (leo-ship-it) [inline boomerang/src/nicp_to_icp.rs:64]: why did you change that? fmt?

File: boomerang/src/nicp_to_icp.rs [hunk 9]
```diff
@@ -97,32 +102,32 @@ pub async fn try_retrieve_icp(target: Principal) -> Result<Nat, BoomerangError>
...
     };
 
-    let icp_balance_e8s: u64 = match icp_client.balance_of(target_account).await {
+    let icp_balance_e8s: u64 = match icp_client.balance_of(boomerang_account).await {
         Ok(balance) => balance.0.try_into().unwrap(),
         Err((code, message)) => {
-            return Err(BoomerangError::GenericError { code, message });
+            return Err(BoomerangError::BalanceOfError(format!(
+                "code: {code} - message: {message}"
+            )));
```
Comment (leo-ship-it) [inline boomerang/src/nicp_to_icp.rs:115]: generic error

File: boomerang/src/nicp_to_icp.rs [hunk 10]
```diff
@@ -97,32 +102,32 @@ pub async fn try_retrieve_icp(target: Principal) -> Result<Nat, BoomerangError>
...
         }
     };
 
-    if icp_balance_e8s < TRANSFER_FEE {
-        return Err(BoomerangError::NotEnoughICP);
+    if icp_balance_e8s == 0 {
```
Comment (leo-ship-it) [inline boomerang/src/nicp_to_icp.rs:119]: Why?

File: boomerang/src/nicp_to_icp.rs [hunk 11]
```diff
@@ -97,32 +102,32 @@ pub async fn try_retrieve_icp(target: Principal) -> Result<Nat, BoomerangError>
...
+        return Err(BoomerangError::IcpNotAvailable);
```
Comment (leo-ship-it) [inline boomerang/src/nicp_to_icp.rs:120]: Keep the precedent error, this one is confusing, what does it mean ICP is not available? Will it become available?

File: boomerang/src/nicp_to_icp.rs [hunk 12]
```diff
@@ -132,12 +137,13 @@ pub async fn try_retrieve_icp(target: Principal) -> Result<Nat, BoomerangError>
             Ok(block_index) => {
                 log!(
                     INFO,
-                    "Transfered ICP for {target} at block index: {block_index}",
+                    "Transfered ICP for {target} at block index: {}",
+                    block_index
```
Comment (leo-ship-it) [inline boomerang/src/nicp_to_icp.rs:141]: why?

File: boomerang/src/state_machine/tests.rs [hunk 13]
```diff
@@ -0,0 +1,73 @@
+use crate::state_machine::{
+    BoomerangSetup, DEFAULT_PRINCIPAL_ID, ONE_MONTH_SECONDS, USER_PRINCIPAL_ID,
```
Comment (leo-ship-it) [inline boomerang/src/state_machine/tests.rs:2]: Boomerang setup feels the same as my code in waterneuron, wouldn't it make sense to reuse it?

File: boomerang/src/state_machine/tests.rs [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+
+    let account_id = boomerang.get_staking_account_id(caller.0);
+
+    assert!(boomerang
+        .icp_transfer(caller.0, None, 1_000 * E8S, account_id)
+        .is_ok());
+
+    assert!(boomerang.notify_icp_deposit(caller.0).is_ok());
+
+    boomerang.advance_time_and_tick(60*60);
```
Comment (leo-ship-it) [inline boomerang/src/state_machine/tests.rs:36]: this is probably not formatted

File: boomerang/src/lib.rs (hunk 2 above)
Comment (uk4z) [reply boomerang/src/lib.rs:38]: "In information technology and computer science, a system is described as stateful if it is designed to remember preceding events or user interactions; the remembered information is called the state of the system. The set of states a system can occupy is known as its state space." wikipedia

Comment (uk4z) [reply boomerang/src/lib.rs:38]: Our canister is not a state hence I came up with another solution to make it work as intended

File: boomerang/src/lib.rs (hunk 3 above)
Comment (uk4z) [reply boomerang/src/lib.rs:83]: I agree. I made the changes necessary to make the test useful again

File: boomerang/src/lib.rs (hunk 4 above)
Comment (uk4z) [reply boomerang/src/lib.rs:87]: The previous configuration was not working for me so I got inspired by the cfg provided in the water_neuron directory

File: boomerang/src/nicp_to_icp.rs (hunk 5 above)
Comment (uk4z) [reply boomerang/src/nicp_to_icp.rs:18]: It is out of consistency. I think it should be either:
```
let boomerang_id = self_canister_id();
    let subaccount = derive_subaccount_staking(target);

    let boomerang_account = Account {
        owner: boomerang_id,
        subaccount: Some(subaccount),
    };
``` 

or
```
    let boomerang_account = Account {
        owner: self_canister_id(),
        subaccount: Some(derive_subaccount_staking(target),
    };
```

But not the in between computation that was there before.

File: boomerang/src/nicp_to_icp.rs (hunk 10 above)
Comment (uk4z) [reply boomerang/src/nicp_to_icp.rs:119]: It is because of the merge of the past repo to waterNeuron. I am changing it back to:
```
if icp_balance_e8s < TRANSFER_FEE {
        return Err(BoomerangError::NotEnoughICP);
        ```

File: boomerang/src/nicp_to_icp.rs (hunk 11 above)
Comment (uk4z) [reply boomerang/src/nicp_to_icp.rs:120]: We shall disagree on this one. To me, NotEnoughICP infers that you expect to have ICP but in the end you don't. 
This error intends to tell the user that the period of 6 months is not over yet and that boomerang did not receive the funds yet. Hence IcpNotAvailable but maybe there is a better name.

File: boomerang/src/state_machine/tests.rs (hunk 13 above)
Comment (uk4z) [reply boomerang/src/state_machine/tests.rs:2]: The struct is similar indeed. But the implementation differs so the methods are specific to boomerang. Plus I like to think of boomerang as an independent module that uses WaterNeuron that's why I did not reuse it.

File: boomerang/src/nicp_to_icp.rs (hunk 6 above)
Comment (uk4z) [reply boomerang/src/nicp_to_icp.rs:33]: done.

File: boomerang/src/icp_to_nicp.rs (hunk 7 above)
Comment (uk4z) [reply boomerang/src/icp_to_nicp.rs:31]: done.

File: boomerang/src/nicp_to_icp.rs (hunk 9 above)
Comment (uk4z) [reply boomerang/src/nicp_to_icp.rs:115]: done.

File: boomerang/src/state_machine/tests.rs (hunk 14 above)
Comment (uk4z) [reply boomerang/src/state_machine/tests.rs:36]: some parts of the code were not but here it was. I formatted everything now.

File: boomerang/src/nicp_to_icp.rs (hunk 8 above)
Comment (uk4z) [reply boomerang/src/nicp_to_icp.rs:64]: done.

File: boomerang/src/nicp_to_icp.rs (hunk 11 above)
Comment (leo-ship-it) [reply boomerang/src/nicp_to_icp.rs:120]: > To me, NotEnoughICP infers that you expect to have ICP but in the end you don't.

No, NotEnoughICP just means that there is not enough ICP to perform the desired action.

> This error intends to tell the user that the period of 6 months is not over yet and that boomerang did not receive the funds yet. Hence IcpNotAvailable but maybe there is a better name.

Then the error is misleading and wrong. The code is absolutely not checking for the desired information. And the name is still bad because ICPNotAvailable could mean that ICP the protocol is not available.

File: boomerang/src/state_machine/tests.rs (hunk 13 above)
Comment (leo-ship-it) [reply boomerang/src/state_machine/tests.rs:2]: Now we have 1k lines of the exact same code. I'm pretty sure we can do better here.
